Add initial integration for `--json=timings`

Kobzol opened the pull request:
### What does this PR try to resolve?

This PR adds initial support into Cargo for JSON timing sections, implemented in rustc in https://github.com/rust-lang/rust/pull/142123. This allows Cargo to read frontend/codegen/linking time from rustc, and thus reporting slightly more detailed data in the `cargo build --timings` output.

Currently this PR modifies Cargo to tell rustc to emit the section messages (`--json=...,timings`), and also renders the sections in the unit table of the unit timings table in the HTML output. As a follow-up, I also want to integrate the individual sections in the actual unit graph.

Note that the JSON timings are currently only supported on the nightly compiler (they are not stabilized). That means two things:
- Cargo has to somehow figure out if it should pass `--json=timings` to `rustc` or not. I don't think that it's worth it to add a new unstable Cargo flag just for this (ofc if you disagree, I'd be happy to implement that), instead I think that Cargo could just detect if the used rustc is nightly, and if yes, pass the flag always. I'm not sure how best to detect that though. So far I just used `gctx.nightly_features_allowed`, but I don't know if it's correct.
- The timings HTML page has to be prepared for the data to be both present and missing. I tried to do that with the `AggregatedSections` enum, and by modifying the table to skip the linking column when no data for it are present. A slight change from before is that both the Frontend and Codegen columns are now always present, while before the Frontend column value had to be "inferred" by the user (by computing `Total` - `Codegen`). I think that's an all-around improvement.

### How to test and review this PR?

You can run e.g. this to generate the timing report with a nightly compiler:
```bash
export RUSTC=`rustup +nightly which rustc`
target/debug/cargo build --timings
```
on some crate, e.g. [ripgrep](https://github.com/BurntSushi/ripgrep).

I have also included example output with and without the section timings: [timings.zip](https://github.com/user-attachments/files/21472709/timings.zip).

epage commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -578,6 +610,46 @@ impl UnitTime {
+
+    fn end_section(&mut self, name: &str, now: f64) {
+        let Some(section) = self.sections.get_mut(name) else {
+            warn!("Compilation section {name} ended, but it has no start recorded");
```
nit: these are generally lower case

epage commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -578,6 +610,46 @@ impl UnitTime {
+            )
+            .is_some()
+        {
+            warn!("Compilation section {name} started more than once");
```
ditto

epage commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -1095,6 +1096,12 @@ fn add_allow_features(build_runner: &BuildRunner<'_, '_>, cmd: &mut ProcessBuild
 ///
 /// [`--error-format`]: https://doc.rust-lang.org/nightly/rustc/command-line-arguments.html#--error-format-control-how-errors-are-produced
 fn add_error_format_and_color(build_runner: &BuildRunner<'_, '_>, cmd: &mut ProcessBuilder) {
+    let enable_timings = build_runner.bcx.gctx.nightly_features_allowed
+        && !build_runner.bcx.build_config.timing_outputs.is_empty();
```
> Cargo has to somehow figure out if it should pass --json=timings to rustc or not. I don't think that it's worth it to add a new unstable Cargo flag just for this (ofc if you disagree, I'd be happy to implement that), instead I think that Cargo could just detect if the used rustc is nightly, and if yes, pass the flag always. I'm not sure how best to detect that though. So far I just used gctx.nightly_features_allowed, but I don't know if it's correct.

Kobzol replied:
Did you mean to reply something here? :sweat_smile: It looks like CI/tests doesn't like using just `nightly_features_allowed`..

epage replied:
I was trying to move your comment in the PR description to an issue on the PR so it has a proper place to be discussed and not be missed.

Kobzol commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -92,6 +92,50 @@ struct UnitTime {
+
+        // Best case: we know all three sections
+        if let (Some(codegen_section), Some(link_section)) = (codegen_section, link_section) {
+            let link_start = link_section.start;
+            let codegen_end = codegen_section.end.unwrap_or(link_start);
+
+            AggregatedSections::Sections {
+                frontend: SectionData {
+                    start: 0.0,
+                    end: codegen_section.start,
+                },
+                codegen: SectionData {
+                    start: codegen_section.start,
+                    end: codegen_end,
+                },
+                linking: SectionData {
+                    start: link_start,
+                    end: link_section.end.unwrap_or(end),
+                },
+            }
+        } else if let Some(rmeta_time) = self.rmeta_time {
+            // We know at least the .rmeta time
+            AggregatedSections::OnlyCodegenDuration {
+                frontend: SectionData {
+                    start: 0.0,
+                    end: rmeta_time,
+                },
+                codegen: SectionData {
+                    start: rmeta_time,
+                    end,
+                },
+            }
+        } else {
+            AggregatedSections::OnlyTotalDuration
+        }
```
Well, that depends on in which direction the rustc feature evolves. In theory we could output anything from rustc and make it fully opaque for downstream tools (including cargo). But then we would probably have to use some hierarchical structure for the sections, which complicates things. And the current build timings output is not really prepared for visualizing such complex data, I think.

So for the first version, I want to only include the three known sections (frontend/backend/linker), as I think that provides the most bang for the buck (e.g. to hint to people that they might try a faster linker, or that the linker is not the bottleneck).

epage replied:
Maybe I'm missing something but I feel like being opaque would be easier to stabilize. 

If we are wanting to be explicit about these names being a part of the format, should we deserialize into an `enum` instead of a `String`?  That also makes it so we don't bury string comparisons in the timings code.

Kobzol replied:
> Maybe I'm missing something but I feel like being opaque would be easier to stabilize.

That might very well be true, if we don't promise the specific contents, it might indeed be easier to stabilize this on the rustc side.

On the other hand, I'm not sure how well we can accommodate arbitrary sections in the timings page. Should we just gather whatever sections we find and then create a single column in the table for each section, in the same order as we saw them in the input? Before we kind of hardcoded frontend/backend from the rmeta time, so I was thinking that we would just extend that with the link time with the section timings.

epage replied:
I'd be fine making the columns dynamic.  Its hard coded now because we have discrete, hard coded pieces of data that cargo collects.

Kobzol commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -549,6 +627,18 @@ impl<'gctx> Timings<'gctx> {
 
     /// Render the table of all units.
     fn write_unit_table(&self, f: &mut impl Write) -> CargoResult<()> {
```
Good catch! Added it.

Kobzol commented:
Ok, so based on feedback I'm fine with moving this to a fully dynamic approach.

Before I do that though, do you have any suggestions on how to detect whether the timings flag should be passed to Cargo? Is there any precedent for doing something in Cargo *only* on the nightly channel, or more precisely only with nightly rustc? Or would you prefer that we add an unstable Cargo flag for this?

Kobzol commented:
Fair enough, I'll add an explicit feature flag then. It's true that if it was implicit, if the rustc format changes in the future, it could break nightly cargo users of `cargo build --timings`. That would be solvable if Cargo was a subtree, but with a submodule it would be quite difficult to ensure that both rustc and cargo change at once.

Kobzol commented:
Ok, I rewrote the PR and updated the PR description. It is now gated behind an unstable `-Zsection-timings` flag, and it renders the sections into the HTML table in a fully dynamic manner.

epage commented on src/cargo/core/compiler/timings.rs:
This commit has the same name as the previous commit

epage commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -90,11 +91,75 @@ struct UnitTime {
+        let end = self.duration;
+
+        if !self.sections.is_empty() {
+            // We have some detailed compilation section timings, so we postprocess them
+            // Since it is possible that we do not have an end timestamp for a given compilation
+            // section, we need to iterate them and if an end is missing, we assign the end of
+            // the section to the start of the following section.
```
Are sections non-overlapping, allowing this to work?

Kobzol replied:
Currently, yes. If it changes in the future, we'd need to of course adapt to that.

epage commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -562,20 +709,53 @@ impl<'gctx> Timings<'gctx> {
       <th>Unit</th>
       <th>Total</th>
-      <th>Codegen</th>
+      {headers}
```
Would it work to split out a refactor commit that adjusts the existing Frontend/Backend logic to this and then add section timings on top?

Kobzol replied:
Hmm, it's all very dependent on one another, but I tried to split it.

Kobzol commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -88,8 +90,9 @@ struct UnitTime {
     unlocked_rmeta_units: Vec<Unit>,
-    /// Individual compilation section durations, gathered from `--json=timings`.
-    sections: HashMap<String, TimingSection>,
+    /// IndexMap is used to keep original insertion order, we want to be able to tell which
+    /// sections were started in which order.
+    sections: IndexMap<String, CompilationSection>,
```
What do you mean by original comment here? :thinking:

epage replied:
Could you please fix that within this commit?

Kobzol commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -480,6 +574,7 @@ impl<'gctx> Timings<'gctx> {
             .enumerate()
             .map(|(i, ut)| (ut.unit.clone(), i))
             .collect();
+
```
Just stylistic I guess, to separate the variable from the item, it looked weird to me:
```rust
let unit_map: HashMap<Unit, usize> = self
    .unit_times
    .iter()
    .enumerate()
    .map(|(i, ut)| (ut.unit.clone(), i))
    .collect();
#[derive(serde::Serialize)]
struct UnitData {
```

epage replied:
Its unrelated to the rest of the commit though.  Please split stylistic changes out to separate commits

epage commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -64,8 +64,9 @@ pub struct Timings<'gctx> {
 }
 
-/// Section of compilation.
-struct TimingSection {
```
Why was this marked as resolved?  I'm not seeing it resolved as requested nor an explanation for why it shouldn't happen

Kobzol commented:
Restructured the commits

epage reviewed (approved):
Thanks!
